File: sdk/derive/address.go
package derive

import (
	"encoding/hex"

	"github.com/bpprotocol/blockparty/sdk/crypto"
)

// Address is a 40-character hex identity address.
type Address string

// BytesToAddress derives the canonical address binding both of an identity's
// post-quantum public keys: hex(Keccak256("v1" || dilithiumPub || kyberPub)[-20:]).
func BytesToAddress(dilithiumPub, kyberPub []byte) Address {
	digest := crypto.Keccak256([]byte("v1"), dilithiumPub, kyberPub)
	return Address(hex.EncodeToString(digest[len(digest)-20:]))
}

// Bytes decodes the address into its 20 raw bytes.
func (a Address) Bytes() ([]byte, error) { return hex.DecodeString(string(a)) }

File: implementations/typescript/packages/sdk/src/derive/address.ts
import { keccak256 } from "../crypto/index.js";
import { bytesToHex, hexToBytes, utf8ToBytes } from "../bytes.js";

/**
 * Derive the canonical 40-char hex address binding both of an identity's
 * post-quantum public keys: hex(keccak256("v1" || mldsaPub || kyberPub)[-20:]).
 */
export function bytesToAddress(mldsaPub: Uint8Array, kyberPub: Uint8Array): string {
  const digest = keccak256(utf8ToBytes("v1"), mldsaPub, kyberPub);
  return bytesToHex(digest.subarray(digest.length - 20));
}

/** Decode an address into its 20 raw bytes. */
export function addressBytes(address: string): Uint8Array {
  return hexToBytes(address);
}
